docs/pages/access-controls/login-rules/reference.mdx

```diff
@@ -0,0 +1,571 @@
+   out.
+
+```yaml
+kind: login_rule
+version: v1
+metadata:
+  name: my_expression_rule
+spec:
+  priority: 0
+
+  traits_map:
+    # the groups trait will be copied unmodified. Do the same for all other
+    # traits which should not be changed, any traits omitted here will *not* be
+    # set for your users and will *not* be used for role mapping.
+    groups:
+      - external["groups"]
+
+    # the logins traits will be set to the username trait converted to
+    # lowercase.
+    logins:
+      - 'strings.lower(external.username)'
+   
+    # the access trait is determined conditionally based on the incoming groups trait.
+    access:
+      - 'ifelse(external.groups.contains("devs"), set("staging"), set())'
+      - 'ifelse(external.groups.contains("admins"), set("staging", "prod"), set())'
+```
+
+### `traits_expression`
+
+Here is an example login rule using the `traits_expression` field implementing
+the same rules as the above example:
+
+```yaml
+kind: login_rule
+version: v1
+metadata:
+  name: my_expression_rule
+spec:
+  priority: 0
+
+  traits_expression: |
+    dict(
+      pair("groups", external.groups),
+      pair("logins", strings.lower(external.username)),
+      pair("access",
+        choose(
+          option(external.groups.contains("devs"), set("staging")),
+          option(external.groups.contains("admins"), set("staging", "prod")),
+          option(true, set()),
+        ),
+      ),
+    )
+```
+
+Every traits expression must return a value of type [`dict`](#dict-type) which will
+be used as the complete set of output traits.
+It is possible to construct a `dict` from scratch as shown above, or you can
+make modifications to the incoming traits stored in the `external` dict like the
+following:
+
+```yaml
+kind: login_rule
+version: v1
+metadata:
+  name: uppercase_logins
+spec:
+  priority: 0
+
+  # This example expression will return all incoming traits unmodified except
+  # for the "logins" trait which will be converted to lowercase.
+  traits_expression: |
+    external.put("logins", strings.lower(external.logins))
+```
+
+## `dict` type
+
+### Description
+
+`dict` is a dictionary type mapping from `string` keys to [`set`](#set-type) values.
+When login rule expressions access input traits with the
+`external.<trait>` or `external[<trait>]` syntax, `external` is a value of type
+`dict`.
+Values of type `dict` can also be constructed and accessed within expressions.
+Expressions used for the `traits_expression` field **must** return a value of
+type `dict`.
+
+### Constructor
+
+#### Signature
+
+```go
+func dict(pairs ...pair) dict
+```
+
+#### Description
+
+The `dict` constructor returns a new dict populated with initial key-value pairs
+from the `pairs` argument.
+Each pair must hold a `string` and a `set`.
+
+#### Arguments
+
+Argument | Type | Description
+-------- | ---- | -----------
+`pairs` | `...pair` | Zero or more key-value pairs to initialize the `dict`.
```

Comment:
Should we include a section for the `pair` type with similar subsections to the `dict` and `set` sections?

Reply:
added type sections for `pair` and `option`. I originally didn't have them since they are pretty limited (no methods) but it's more consistent this way I think